### ISSUE TITLE
File handles left open on program termination

### ISSUE THREAD

**Issue by WasabiFan** [ev3dev/ev3dev-lang-python#433]
Currently, we store handles to the attribute files we open for quick access. This means that we get a bunch of warning spew like this from the tests:

```
/opt/python/3.4.6/lib/python3.4/unittest/case.py:580: ResourceWarning: unclosed file <_io.FileIO name='/home/travis/build/ev3dev/ev3dev-lang-python/tests/fake-sys/arena/lego-sensor/sensor0/address' mode='rb'>
```

It would be preferable if we kept track of the handles and closed them explicitly in `__del__` rather than leaving them for the interpreter to (hopefully) clean up. This would normally be pretty easy, but all of our calls to read from attributes in the device subclasses look like this:

```python
self._address, value = self.get_attr_string(self._address, 'address')
```

In other words, there is no centralized location where we store the file handles. What do others think? I would be in favor of moving the cache a level down into the device base class.

There is the alternative option of configuring the unit tests to ignore warnings... but that feels a bit wrong.

Side-note: Autogen would come in really handy in this case! :laughing:

**Comment by ddemidov** [comment]
I think we discussed this somewhere, and the problem with file handles was that python does not have proper class destructors. For example, see https://stackoverflow.com/questions/35488705/is-relying-on-del-for-cleanup-in-python-unreliable.

We could probably come up with an interface like
```py
with LargeMotor(OUTPUT_A) as m, ColorSensor() as cs:
    # do stuff with m and cs, clean up on exit
```
but that would be inconvenient for use if one wants to store a device as a class field. My understanding is that what we have now is considered acceptable in python, but I am no expert here.

I am not sure what the problem with centralized file caching was, but we had it implemented at some point and moved away for some reason. :ghost:

**Comment by ddemidov** [comment]
#109, #96

**Comment by ddemidov** [comment]
#162 (4a873844cdeff38ebc5d870f8c8bdfedefa2e8e3)

**Comment by WasabiFan** [comment]
My recollection was that we had a `__del__` implementation which would stop motors on termination, and that had some obvious issues, especially because we were attempting to go through the file file write process. I am guessing that closing file handles wouldn't be problematic, but I'm not sure. My thinking is that closing file handles explicitly is nice to have for sanity, but it doesn't affect program behavior if that code doesn't run when we expected it to. Meanwhile, I'm having a tough time deciphering why removed the file cache...

**Comment by dlech** [comment]
> Meanwhile, I'm having a tough time deciphering why removed the file cache...

https://github.com/ev3dev/ev3dev-lang-python/issues/155#issuecomment-209199499

**Comment by dlech** [comment]
> My recollection was that we had a `__del__` implementation which would stop motors on termination

That was the intention, but `__del__` only runs on garbage collection, so it practice, it is rarely, if ever, called since the motor objects stay alive for the entire program. So, it would be "correct" to close any open file handles in `__del__`, but it most likely won't fix your unit test.

**Comment by dwalton76** [comment]
What if we close them in `__del__` and have the test cases call `gc.collect()`?

**Comment by dlech** [comment]
Even if you manually run the garbage collector, there is no guarantee that `__del__` will be called. It could only happen if there are no references to the object remaining. If you want to actually close all file handles, then you should implement a `close()` method that does so.
